newhelm/typed_data.py

```diff
@@ -0,0 +1,32 @@
+from typing import Any, Dict, Type, TypeVar
+from typing_extensions import Self
+from pydantic import BaseModel
+
+
+_BaseModelType = TypeVar("_BaseModelType", bound=BaseModel)
+
+
+class TypedData(BaseModel):
+    """This is a generic container that allows Pydantic to do polymorphic serialization.
+
+    This is useful in situations where you have an unknown set of classes that could be
+    used in a particular field.
+    """
+
+    type: str
+    data: Dict[str, Any]
+
+    @classmethod
+    def from_instance(cls, obj: _BaseModelType) -> Self:
+        """Convert the object into a TypedData instance."""
+        return cls(type=TypedData._get_type(obj.__class__), data=obj.model_dump())
+
+    def to_instance(self, cls: Type[_BaseModelType]) -> _BaseModelType:
+        """Convert this data back into its original type."""
+        cls_type = TypedData._get_type(cls)
+        assert cls_type == self.type, f"Cannot convert {self.type} to {cls_type}."
+        return cls.model_validate(self.data)
+
+    @staticmethod
```

Comment:
Did we mean to use `@classmethod` here for the `cls` argument?

Reply:
Oh, this is just my confusion with the naming of `cls`. Maybe we change this to something else to prevent that confusion?

Reply:
Updated to `instance_cls` here and in a few other methods.

Reply:
Add type signature to method?

Reply:
Done.